---
title: "6.1 For Loops - Tutorial"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error = TRUE, warning = TRUE)

library(ottr)
library(testthat)
library(dplyr)
```

---

At the end of this tutorial you should be able to:

* Write a for loop with the correct syntax
* Loop through vectors
* Loop through vector indices and data frame rows
* Modify data structures from within a for loop
* Determine how to implement a for loop for specific applications

**NOTE: the dplyr package has already been loaded for you**

---

## 6.1.1: Introduction to For Loops

---

Consider the following for loop:

```{r}
for (num in c(3, 9, 27, 81, 243)) {
  print(num / 3)
}
```

Assign the elements that the for loop above will iterate through to a variable called "elements".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W6.1.1.a.R")
```
**Check your answer by running the above block** 

Assign the number of iterations in the for loop above to a variable called "iterations".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W6.1.1.b.R")
```
**Check your answer by running the above block** 

If during the 3rd iteration you looked at the value of the variable `num`, what would it be? Assign your answer to a variable called "num_3rd_iter".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W6.1.1.c.R")
```
**Check your answer by running the above block** 

---

## 6.1.2: For Loop Techniques

---

Write a for loop that replaces each number in `num_vect` with the square root of the number.
```{r}
num_vect <- c(1, 4, 9, 16, 25, 36, 49, 64, 81, 100)
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W6.1.2.a.R")
```
**Check your answer by running the above block** 

---

Below is a vector of the type of cancer for each of 20 patients (`patient_cancer_types`), and a vector of all unique the cancer types (`cancer_types`). Using a for loop, fill the empty vector `patients_per_cancer_type` so that it has one element for each cancer type. The value at each element should be the number of patients with the cancer type and the element should be named with the cancer type.

Hint: to count the number of values in a vector that match a specific value you can use:

    length(which(vector == value))
```{r}
patient_cancer_types <- c("colon", "breast", "pancreatic", "gastric",  "gastric", "breast",  "pancreatic", "gastric",  "gastric",  "gastric", "colon", "colon", "breast", "pancreatic", "pancreatic", "colon", "gastric", "colon", "gastric", "breast")

cancer_types <- c("breast", "colon", "gastric", "pancreatic")

patients_per_cancer_type <- c()

# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W6.1.2.b.R")
```
**Check your answer by running the above block** 

---

The Fibonacci sequence is a sequence of numbers in which each number is the sum of the previous two numbers. The following vector contains the first 4 numbers in the Fibonacci Sequence. As you can see, the 3rd number in the sequence is 1, which is the sum of the prior two numbers (0 + 1), and the 4th number in the sequence is 2, which is the sum of the prior two numbers (1 + 1).

```{r}
fibonacci_sequence <- c(0, 1, 1, 2)
```

Write a for loop to add the next 8 numbers to the `fibonacci_sequence` vector. Hint: this will mean you need to add numbers at the 5th - 12th indices.
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W6.1.2.c.R")
```
**Check your answer by running the above block** 

---

## 6.1.3: For Loop Applications

---

For the rest of this tutorial we will used data downloaded from ClinVar (https://www.ncbi.nlm.nih.gov/clinvar/), a public database of human gene variants and their associated phenotypes.

The dataset we will use includes all the variants on ClinVar associated with the following 5 diseases:

- Cystic fibrosis
- Duchenne muscular dystrophy
- Glaucoma
- Phenylketonuria
- Type 2 diabetes mellitus

Upload and view the data set with the code below. A vector called `diseases` has also been created for you to use.

```{r}
diseases <- c("cystic.fibrosis", "duchenne.muscular.dystrophy", "glaucoma", "pheynlketonuria", "type.2.diabetes")
clinvar_data <- read.csv("/home/jovyan/Week.6/6.1.For.Loops/Tutorial.6.1/clinvar_data.csv",
                         stringsAsFactors = FALSE)
clinvar_data
```

Each row in the data frame contains one variant with the following information:

- `chromosome`: chromosome the variant is on
- `position`: exact base position of the variant
- `disease`: the disease the patient(s) with the variant has(have)
- `gene`: the gene in which the variant is located
- `clinical.significance`: defines whether the variant is pathogenic or not, possible values are: "Pathogenic", "Likely_pathogenic", "Benign", "Likely_benign", "Uncertain", "Other"

---

A new column `pathogenic` has been added to the `clinvar_data` data frame below to help us more easily analyze the variants that are pathogenic or likely pathogenic. Fill the column by looping through each row in the data frame and setting the value for that row in the `pathogenic` column to `TRUE` if the `clinical.significance` in that row is "Pathogenic" or "Likely_pathogenic", or to `FALSE` if it has any other value.
```{r}
clinvar_data$pathogenic <- NA
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W6.1.3.a.R")
```
**Check your answer by running the above block** 

What percentage of variants for each disease is pathogenic or likely pathogenic? 

To answer this question, write a for loop to fill the vector below `perc_pathogenic_variants`. It should contain the percentage of pathogenic or likely pathogenic variants (use your new column!) for each disease. Each vector element should be named with the disease name. 
```{r}
perc_pathogenic_variants <- c()
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W6.1.3.b.R")
```
**Check your answer by running the above block** 

---

Before you complete the next question, let's learn the function `n_distinct()`. It can be applied to a vector or to a data frame column, and it will return the number of distinct/unique values. For example:

```{r}
# There are 48 variants in the MYOC gene
# These variants have 3 distinct types of clinical significance
n_distinct(clinvar_data %>% filter(gene == "MYOC") %>% select(clinical.significance))
```

For each disease, how many unique genes have a pathogenic variant associated with the disease? Which disease can be caused by the most genes (according to this data set)?

Write a for loop to fill the vector below `num_genes_with_pathogenic_variants`. It should contain the number of distinct genes that contain pathogenic/likely pathogenic variants for each disease. Each vector element should be named with the disease name. Assign the name of the disease with the most distinct genes with pathogenic variants to a variable named "disease_with_most_pathogenic_genes".
```{r}
num_genes_with_pathogenic_variants <- c()
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W6.1.3.c.R")
```
**Check your answer by running the above block** 

---
