---
title: "6.3 Loop Applications"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---

## 6.3.1 Nested Loops

---

### Nested For Loops

Often we need to use nested for loops, which allows us to loop through each combination of items in two sets:

    for (i in items_1) {
      for (j in items_2) {
        ... code block ...
      }
    }

In this representation of nested loops, we go through each item `i` in `items_1` and each item `j` in `items_2`. The sets, `items_1` and `items_2`, can be the same set of items or two different sets of items, but the iteration variable names, `i` and `j`, must always be different from one another. 

In total, the number of iterations of two nested for loops is the product of the length of each set of items: `length(items_1) * length(items_2)`. If you nested a third for loop, it would the product of all three sets: `length(items_1) * length(items_2) * length(items_3)`.

If `items_1` contains three elements and `items_2` contains three elements, there will be a total of 3 x 3 = 9 iterations.

To better understand how nested for loops work, let's look at an example. The outer loop iterates through a vector containing the numbers 1 to 3, and the inner loop iterates through a vector that contains numbers 5 and 10. At each iteration, `num1` and `num2` are multiplied and the equation along with the result is printed to the screen:

```{r}
for (num1 in 1:3) {
  for (num2 in c(5, 10)) {
    product <- num1 * num2
    print(paste(num1, "x", num2, "=", product))
  }
}
```

In the first iteration of the outer loop, `num1` is equal to 1. While `num1` is equal to 1, the inner loop runs. In the first iteration of the inner loop, `num2` is equal to 5, and then `num2` is equal to `10` in the second iteration of the inner loop. When this inner loop finishes, the outer loop will move on to the next value of `num1`.

---

### Looping through Matrices

One of the key uses of nested for loops is looping through rows and columns in a matrix. Matrices can be looped through using row and column indices:
    
    for (row in 1:nrow(matrix_name)) {
      for (col in 1:ncol(matrix_name)) {
        ... matrix_name[row, col] ...
      }
    }

Or with row and column names:

    for (row_name in rownames(matrix_name)) {
      for (col_name in colnames(matrix_name)) {
        ... matrix_name[row_name, col_name] ...
      }
    }

Each of the above sets of loops will loop through each row, and then through each column. 

Let's look at an example. Below is a matrix containing the value NA for each element:

```{r}
my_matrix <- matrix(NA, nrow=3, ncol=2)
my_matrix
```

Let's loop through each row and column in the matrix and change the value of the element so that it is the product of the row and column indices, and print after each inner loop iteration to monitor our progress:

```{r}
for (row_num in 1:nrow(my_matrix)) {
  for(col_num in 1:ncol(my_matrix)) {
    my_matrix[row_num,col_num] <- row_num * col_num
    print(my_matrix)
  }
}
```

If we look at the output, we can see that in the first iteration, the value at row 1, column 1 was changed to 1. In the second iteration, the value at row 1, column 2 was changed to 2. If you continue to look at output you can see that each row was completed before the next row began, because the row indices were in the outer loop.

--- 

Let's again look at examples using brain tissue expression data sourced from the Human Protein Atlas (https://www.proteinatlas.org/humanproteome/brain):

```{r}
brain_expression <- read.csv("/home/jovyan/Week.6/6.3.Applying.Loops/brain_expression.csv", stringsAsFactors = FALSE)
brain_expression
```

As a refresher, each of the rows represents one of 15,271 genes. The first column is Ensembl Gene ID of each gene, which is a unique gene identifier. We will learn more about Ensembl Gene IDs later on in the course. The second to eleventh columns contain normalized gene expression values for 10 different brain tissues. 

The names of these tissues (which are the column names for columns 2-11) are saved in the vector `brain_tissues`:

```{r}
brain_tissues <- colnames(brain_expression)[2:11]
brain_tissues
```

Let's create a matrix of the correlation of expression between all 10 brain tissues. We will start by creating an empty matrix with 10 rows and 10 columns, which each value set to NA. The names of the 10 tissues are used to name the rows and columns:

```{r}
tissue_expr_cor <- matrix(NA, nrow=10, ncol=10)
rownames(tissue_expr_cor) <- brain_tissues
colnames(tissue_expr_cor) <- brain_tissues
tissue_expr_cor
```

In supplement 6.1 we calculated the correlation in gene expression between the amygdala and each of the other brain tissues (including itself). Let's repeat this, but for all pairings of all 10 tissues. We can use nested for loops: first looping through the row names of the matrix and then through the column names. Within each iteration we can calculate the expression correlation between the two tissues, round it, and add the value to our `tissue_expr_cor` matrix in the appropriate spot:

```{r}
for (rowname in rownames(tissue_expr_cor)) {
  for (colname in colnames(tissue_expr_cor)) {
    r <- cor(brain_expression[, rowname], brain_expression[, colname])
    r <- round(r, 2)
    tissue_expr_cor[rowname, colname] <- r
  }
}
tissue_expr_cor
```

---

### Nesting For and While Loops

We can also nest while loops within while loops, for loops within while loops, and while loops within for loops. Let's look at an example of the latter. 

In supplement 6.2 we calculated the minimum number of genes that accounted for 50% of the RNA expression in the midbrain, and used that value to determine the percentage of genes that accounted for 50% of the RNA. How could we calculate the percentage of genes that accounted for 50% of the expression for all 10 tissues? 

We could use a for loop to go through the tissues, and a while loop to get the number of genes. We will start with an empty vector "percent_genes_for_half_RNA" to store our final gene percentages for each tissue. Within the for loop, we will sort expression values for the current tissue and determine half of the expression. Using the same nested while loop that we used in supplement 6.2, we will calculate the number of genes it takes to make up 50% of the expression for the tissue. After the while loop terminates, we can calculate and add the percentage to our results vector within the for loop before starting on the next tissue iteration.

In previous nested loops we did not include any code in between the declaration of the outer and inner loops, but we can write code that occurs for each item in the outer loop before running the inner loop.

```{r}
percent_genes_for_half_RNA <- c()
for (tissue in brain_tissues) {
  # sort gene expression values for the current tissue
  expr <- sort(brain_expression[, tissue], decreasing = TRUE)
  # calculate what half of the total expression is
  half_expr <- sum(expr)/2
  # prepare variables for the inner while loop
  cumulative_expr <- 0
  index <- 0
  while (cumulative_expr <= half_expr) {
    # update index
    index <- index + 1
    # add current gene expression value to cumulative total for the current tissue
    cumulative_expr <- cumulative_expr + expr[index]
  }
  # add % of genes that make up half of the expression to the tissue vector
  percent_genes_for_half_RNA[tissue] <- 100*index/length(expr)
}
percent_genes_for_half_RNA
```

When we look at the output we can see that for all of the tissues, around 18-21% of genes account for half of all expression.

---

## 6.3.2 Vectorization

---

### Non-vectorized vs Vectorized Code

In supplement 6.1 we looked at a real-life example that is similar to a for loop: a scientist is working in the lab with six samples, and for each sample the scientist needs to add 10 uL of lysis buffer and then place the sample on ice:

    For X in sample 1, sample 2, sample 3, sample 4, sample 5, sample 6:
      - Add 10 uL of lysis buffer to X
      - Place X on ice

In this example, the biologist does one sample and then moves onto the next. But what if all the tasks could be performed simultaneously? This could be done with:
- Six scientists performing the task, each performing the task for one sample
- One scientist using a multichannel pipette to add the buffer to all of the samples at the same time

In R, speeding up a task like this is called *vectorization*. Most functions that we have learned and applied so far are vectorized: they perform their task simultaneously for all of the items they are provided. 

Vectorized functions we have learned so far include `log()`, `round()`, `sqrt()`, `gsub()`, and `grepl()`.

Vectorized code is much more efficient and concise than non-vectorized code. Let's compare non-vectorized vs vectorized code to take the natural log of the numbers 1 through 4, adding the results to a vector named "y".

With a for loop (non-vectorized), we can do the following:

```{r}
y <- c()
for (x in c(1:4)) {
  y <- c(y, log(x))
}
y
```

The `log()` function in R, however is vectorized. If you run it on a vector, it will be applied to each item in the vector:

```{r}
y <- log(1:4)
y
```

It is clear that this code is more concise, but it is also faster? To time how long a piece of code takes to run we can use the `Sys.time()` function, which outputs the current time when it is executed. By running `Sys.time()` before and after a piece of code and saving the values to variables, we can calculate the difference between the start and end times to find out how long it took for the code to run.

First, let's calculate how long the non-vectorized code (for loop) takes to run. To get accurate times, be sure to run the entire code block at once (instead of line-by-line):

```{r}
start_time <- Sys.time()
y <- c()
for (x in c(1:4)) {
  y <- c(y, log(x))
}
end_time <- Sys.time()

non_vectorized_code_runtime <- end_time - start_time
"Runtime:"
non_vectorized_code_runtime
```

Now let's calculate the runtime for vectorized code:

```{r}
start_time <- Sys.time()
y <- log(1:4)
end_time <- Sys.time()

vectorized_code_runtime <- end_time - start_time
"Runtime:"
vectorized_code_runtime
```

Your runtimes may differ, but the runtimes during the writing of this supplement were:
- Non-vectorized code runtime: 0.4954598 seconds
- Vectorized code runtime: 0.1537697 seconds

The vectorized code ran about 3.22x faster. This may not seem important since both run times are less than 1 second, but when we are working with much larger vectors, this becomes very important.

Let's look at the same example, but take the log of the numbers 1 through 100,000 (this code block may take a while to run!):

```{r}
# Non-vectorized code
start_time <- Sys.time()
y <- c()
for (x in c(1:100000)) {
  y <- c(y, log(x))
}
end_time <- Sys.time()
non_vectorized_code_runtime <- end_time - start_time

# Vectorized code
start_time <- Sys.time()
y <- log(1:100000)
end_time <- Sys.time()
vectorized_code_runtime <- end_time - start_time

# Runtimes
"Non-vectorized code runtime:"
non_vectorized_code_runtime
"Vectorized code runtime:"
vectorized_code_runtime
"Vectorized code is X times faster:"
as.numeric(non_vectorized_code_runtime)/as.numeric(vectorized_code_runtime)
```

Using vectorized functions is extremely helpful when we are working with large datasets, and most biological datasets are indeed very large.

---

### Vectorizing Your Code

How do you know when to use a loop or vectorized code? Here is a good guideline:

| Do you know how many times the task needs to be repeated? 
  |--- NO  -> use a while loop
  |--- YES -> use a for loop or vectorized code
       | Is there an R function that can perform some or all of the task?
         |--- NO  -> use a for loop
         |--- YES -> use a vectorized function
      
There is also a way to write your own vectorized functions. While we will not be focusing on the method in this course, it will be briefly covered in supplement 7.1.

Let's look at some practical examples in which we can use vectorized code using the brain tissue expression data.

Calculate the mean expression level for each brain tissue. How many times do we need to perform this task? 10 times, once per tissue. Therefore, we should use either a for loop or a vectorized function. First let's see how we would do this with a for loop:

```{r}
tissue_expr_means <- c()
for (tissue in brain_tissues) {
  tissue_mean <- mean(brain_expression[, tissue])
  tissue_expr_means <- c(tissue_expr_means, tissue_mean)
}
tissue_expr_means
```

The for loop works, however the function `colMeans()` calculates the mean for each column in a data frame or matrix automatically. We can vectorize the above code by using the `colMeans()` function instead:

```{r}
tissue_expr_means <- colMeans(brain_expression[,2:11])
tissue_expr_means
```

In fact, the nested loops we used earlier to calculate the correlation between expression values for each pair of tissues can also be vectorized using the `cor()` function:

```{r}
cor(brain_expression[,2:11])
```

Other useful vectorized functions that are frequently used with data frames include `colSums()` and `rowMeans()`.

---

### Code Vectorization Examples

Let's look at some more examples of vectorization using a new data set that contains genes and information about which chromosome each gene is on, as well as the strandedness. The data frame contains the following columns:
`Gene.ID` - Ensembl unique gene identifier
`Chromosome` - the name of the chromosome each gene is on
`Strand` - whether the gene is on the positive strand (+) or the negative strand (-)

```{r}
genes_and_chromosomes <- read.csv("/home/jovyan/Week.6/6.3.Applying.Loops/genes_and_chromosomes.csv", stringsAsFactors = FALSE)
genes_and_chromosomes
```

To create a vector containing the number of genes on chromosome X or Y, we could write a for loop to go through each row and check if it is on one of these two chromosomes:

```{r}
# Initiate an empty vector to store the count of genes
num_genes_X_Y <- 0
for (gene in 1:nrow(genes_and_chromosomes)) {
  # If statement to check if the gene is on chromosome X or Y
  if (genes_and_chromosomes[gene, "Chromosome"] == "X" | genes_and_chromosomes[gene, "Chromosome"] == "Y") {
    # If the gene is on chr X or Y, add to the count of genes
    num_genes_X_Y <- num_genes_X_Y + 1
  }
}
num_genes_X_Y
```

However, it can also easily be vectorized by using the functions `nrow()` and `filter()` (from the `dplyr` package):

```{r}
library(dplyr, warn.conflicts = F)
nrow(filter(genes_and_chromosomes, Chromosome == "X" | Chromosome == "Y"))
```

For each of the chromosomes in our data frame, how many genes does it have on the positive strain? Using nested for loops we can create a list named "chromosome_pos_genes", with each list element having the name of a chromosome and containing the number of positive-strand genes on that chromosome. 

Note that the `unique()` function returns the unique values in a vector. For example:

```{r}
unique(genes_and_chromosomes$Chromosome)
```

```{r}
# Write and execute your code here
chromosome_pos_genes <- list()
for (chr in unique(genes_and_chromosomes$Chromosome)) {
  # Filter dataset so it only contains genes on the chromosome being iterated
  chr_rows <- filter(genes_and_chromosomes, Chromosome == chr)
  # Initiate empty vector to store the count of genes
  positive_strand_genes <- 0
  # Iterate through each row in the filtered dataset
  for (row in 1:nrow(chr_rows)) {
    # Check if the gene in the row being iterated is on the + strand
    if (chr_rows$Strand[row] == "+") {
      # If it is, increase our gene count vector for this chromosome
      positive_strand_genes <- positive_strand_genes + 1
    }
  }
  # Add the gene count to our results list in the appropriate location
  chromosome_pos_genes[[chr]] <- positive_strand_genes
}
chromosome_pos_genes
```

The inner loop in the above code can be vectorized using the same technique we used when getting the number of genes on the X and Y chromosomes: `nrow(filter())`.

```{r}
# Write and execute your code here
chromosome_pos_genes <- list()
for (chr in unique(genes_and_chromosomes$Chromosome)) {
  # Filter dataset so it only contains genes on the chromosome being iterated
  chr_rows <- filter(genes_and_chromosomes, Chromosome == chr)
  # Further filter dataset so it only contains genes on the positive strand
  # and then get the number of rows (gene count)
  positive_strand_genes <- nrow(filter(chr_rows, Strand == "+")) 
  # Add the gene count to our results list in the appropriate location
  chromosome_pos_genes[[chr]] <- positive_strand_genes
}
chromosome_pos_genes
```

---