---
title: "Week 6 Assignment"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error = TRUE, warning = TRUE)

library(ottr)
library(testthat)
library(dplyr)
```
**Run the above block to load required packages**

---

## Overview

---

This assignment covers the content in week 6 lectures and tutorials. 

Although it is not required to answer any of the questions, the dplyr package has been loaded should you wish to use it.

#### Mark breakdown
- Question 1 (10 marks)
- Question 2 (6 marks)
- Question 3 (4 marks)

---

## Background

---

The Genotype-Tissue Expression (GTEx) project has collected samples from 54 different human tissues from ~1000 non-diseased individuals and performed RNA-seq. The GTEx portal (https://gtexportal.org/) contains publicly available expression data for all 54 human tissues. 

In this assignment we will be working with the GTEx expression data. The data frame below contains one row for each of the 56,200 genes with expression data, and 56 columns. 

```{r}
GTEx_data <- read.table("/home/jovyan/Week.6/Week.6.Assignment/GTEx_expression_data.txt", header=TRUE, stringsAsFactors = FALSE, sep="\t")
GTEx_data
```

Description of data frame columns:

*gene.ID*               Ensembl unique identifier for each gene           
*gene.name*             The gene name (ex. "TP53")

Columns 3 to 56 contain the expression data for each of the 54 tissue types, the names of which are also contained in the vector `tissues` below:

```{r}
tissues <- colnames(GTEx_data)[3:56]
tissues
```

---

## QUESTIONS

---

### Question 1 

#### a.
(2 marks)
Use a for loop to create a vector named "median_tissue_expression" where each element is named with the name of the tissue, and the value is the median expression for that tissue.
```{r Assignment.W7.1.a}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W6.1.a.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### b.
(3 marks)
Use a for loop to add a new column "highest.tissue.exp" to `GTEx_data` which contains, for each row, the name of the tissue in which the gene is the most highly expressed. If the gene is not expressed in any tissues (0 for all tissues) put "none" instead of a tissue name.

Hints: 
- The function `which.max()` will return the index of the highest number in a vector.
- The vector `tissues` has, in order, the column names of the GTEx columns that contain the expression data.
```{r Assignment.W7.1.b}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W6.1.b.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### c.
(2 marks)
The file "gene.types.txt" contains gene IDs and the type of each gene. This can be one of three values:
- "coding" : the gene codes for a protein
- "nonCoding" : the gene does not code for a protein. This includes lncRNAs (long non-coding RNAs), rRNAs (ribosomal RNAs), and tRNAs (transfer RNAs)
- "pseudo" : the gene is a pseudogene, a nonfunctional DNA segment that resembles a gene. It does not code for a protein but can still be transcribed

```{r Assignment.W7.1.c}
gene_types <- read.table("/home/jovyan/Week.6/Week.6.Assignment/gene.types.txt", stringsAsFactors = FALSE, sep="\t", header=TRUE)
gene_types
```
Use a for loop to add a new column "gene.type" to `GTEx_data` which contains for each row, the type of gene it is ("coding", "nonCoding", "pseudo"). If the type for the gene is not in `gene_types`, instead put "unknown".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W6.1.c.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### d.
(3 marks)
Below is a matrix called `tissue_exp_gene_type` with 54 columns (one per tissue) and 3 rows (one per gene type). The columns have been named accordingly and all values in the matrix are `NA`. Use nested for loops to fill the matrix with the number of each of the gene types that are expressed > 1 in each tissue type.

For example, in liver tissue, 11645 coding genes have an expression value > 1, 1271 nonCoding genes have an expression value > 1, and 516 pseudo genes have an expression value > 1. Use this to check your work.
```{r Assignment.W7.1.d}
tissue_exp_gene_type <- matrix(data=NA, nrow=3, ncol=54)
colnames(tissue_exp_gene_type) <- tissues
rownames(tissue_exp_gene_type) <- c("coding", "nonCoding", "pseudo")
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W6.1.d.R")
```
**IGNORE the above block: test results will not appear in assignments** 

---

### Question 2

#### a. 
(3 marks)
The data frame below `GTEx_data_Lymphocytes_desc` contains the same data as `GTEx_data` but the genes are arranged in descending order according to their expression in lymphocytes. This means that the gene most highly expressed in lymphocytes is in row 1, the second most highly expressed gene in lymphocytes is in row 2, etc.

```{r Assignment.W7.2.a}
GTEx_data_Lymphocytes_desc <- arrange(GTEx_data, desc(Lymphocytes))
GTEx_data_Lymphocytes_desc
```
Use a while loop to identify the gene names of the 10 most highly expressed genes in lymphocytes ("Lymphocytes") that are NOT expressed in whole blood ("Whole.Blood"). The definition of NOT expressed is an expression value < 1. Assign the names of these genes to a vector called "q.2.a". 
Use the `GTEx_data_Lymphocytes_desc` data frame to answer this question.
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W6.2.a.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### b.
(3 marks)
Use a while loop to identify the name of the most highly expressed gene in lymphocytes that is not expressed in any other tissue type (not expressed = expressed below 1). Assign the name of this gene to a variable named "q.2.b".

Hint: "Lymphocytes" is column 25, so to get all the columns for the other 53 tissues you can use the code: `GTEx_data_Lymphocytes_desc[, c(3:24, 26:56)]`
```{r Assignment.W7.2.b}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W6.2.b.R")
```
**IGNORE the above block: test results will not appear in assignments** 

---

### Question 3

For this question you can use loops or vectorized methods (your mark will not be affected by the method you choose). It only matters that you get the correct answer.

#### a.
(1 mark)
Use either type of loop or vectorized code to create a vector named "mean_tissue_expression". Each element should be named with the name of the tissue, and the value should be the mean tissue expression.
```{r Assignment.W7.3.a}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W6.3.a.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### b.
(2 marks)
Use either type of loop or vectorized code to identify the gene with the highest expression in the kidney cortex ("Kidney.Cortex") that is not expressed below 1 in the kidney medulla ("Kidney.Medulla"). Assign the name of the gene to a variable called "q.3.b".
```{r Assignment.W7.3.b}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W6.3.b.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### c.
(1 mark)
Use either type of loop or vectorized code to determine the mean and standard deviation of the expression of pseudogenes in the cerebellum ("Brain.Cerebellum").
Put the mean and standard deviation into a vector named "q.3.c" (mean should be at index 1, standard deviation at index 2, and elements should NOT have names). 
```{r Assignment.W7.3.c}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W6.3.c.R")
```
**IGNORE the above block: test results will not appear in assignments** 

---